lib/Twig/Extension/Core.php

```diff
@@ -643,16 +646,25 @@ function twig_last(Twig_Environment $env, $item)
  *
  * @param array  $value An array
  * @param string $glue  The separator
+ * @param string $and   The separator for the last pair
  *
  * @return string The concatenated string
  */
-function twig_join_filter($value, $glue = '')
+function twig_join_filter($value, $glue = '', $and = null)
 {
     if ($value instanceof Traversable) {
         $value = iterator_to_array($value, false);
     }
 
-    return implode($glue, (array) $value);
+    if( empty( (array) $value ) || is_null( $and ) || $and === $glue ) {
```

Comment:
The null check should use `null === $and` rather than `is_null()` according to the coding standards